src/localization/gps_linearization.py

```diff
@@ -28,59 +31,90 @@ class GPSLinearization:
     ref_lon: float
     ref_alt: float
 
-    # frame configuration
+    # frame/publishing configuration
+    publish_tf: bool
     use_odom: bool
     world_frame: str
     odom_frame: str
     rover_frame: str
 
-    def __init__(self):
-        # init to zero pose
-        self.pose = SE3()
-
-        rospy.Subscriber("gps/fix", NavSatFix, self.gps_callback)
-        rospy.Subscriber("imu/data", ImuAndMag, self.imu_callback)
-
-        self.tf_broadcaster = tf2_ros.TransformBroadcaster()
-        self.tf_buffer = tf2_ros.Buffer()
-        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer)
+    # covariance config
+    use_dop_cov: bool
+    config_gps_covariance: np.ndarray
 
+    def __init__(self):
         # read required parameters, if they don't exist an error will be thrown
         self.ref_lat = rospy.get_param("gps_linearization/reference_point_latitude")
         self.ref_lon = rospy.get_param("gps_linearization/reference_point_longitude")
         self.ref_alt = rospy.get_param("gps_linearization/reference_point_altitude")
 
-        self.use_odom = rospy.get_param("gps_linearization/use_odom_frame")
-        self.world_frame = rospy.get_param("gps_linearization/world_frame")
-        self.odom_frame = rospy.get_param("gps_linearization/odom_frame")
-        self.rover_frame = rospy.get_param("gps_linearization/rover_frame")
+        self.publish_tf = rospy.get_param("gps_linearization/publish_tf")
+        self.use_odom = rospy.get_param("use_odom_frame")
+
+        self.world_frame = rospy.get_param("world_frame")
+        self.odom_frame = rospy.get_param("odom_frame")
+        self.rover_frame = rospy.get_param("rover_frame")
+
+        self.use_dop_cov = rospy.get_param("global_ekf/use_gps_dop_covariance")
+        self.config_gps_covariance = np.array(rospy.get_param("global_ekf/gps_covariance", None))
+
+        # init to zero pose and either zero covariance or configured covariance
+        self.pose = SE3()
+        self.covariance = np.zeros((6, 6))
+        if not self.use_dop_cov:
+            self.covariance[:3, :3] = self.config_gps_covariance
```

Comment:
I'm really confused by this, is this not the post-filter output? The covariance of the pose should be the filter's state covariance. The default initialization to zero with no else here also seems wrong.

Reply:
This isn't post filter output, this is a result of how I'm fusing GPS and IMU pose into the EKF. The GPS accepts a PoseWithCovarianceStamped message as sensor measurement input and uses the included covariance as the sensor measurement covariance. So what I'm doing here is constructing a pose from the raw GPS position and IMU orientation, and then constructing the corresponding covariance matrix from the individual GPS position and IMU orientation covariance matrices.

Reply:
ok got it, but why is there no else for when use_dop_cov is set?

Reply:
nevermind, I see it on line 139 in the gps callback